### DISCUSSION TITLE
Does Uno under any circumstances require events to be manually unsubscribed?

### DISCUSSION THREAD

**Discussion by abenadar** [unoplatform/uno#14240] [Q&A] [answered]
According to the [Windows Apps documentation](https://learn.microsoft.com/en-us/windows/apps/develop/win2d/avoiding-memory-leaks) there are certain cases in which events must be unsubscribed in the Unloaded event to avoid memory leaks.

I am not using any of these controls but I am wondering whether Uno has any pitfalls like this as well. More sepcifically I'd like to know: 

- Are events subscribed to in Xaml ever required to be unsubscribed?
- Are events subscribed to in code-behind ever required to be unsubscribed?
- Is there anything else here to be cautious of?

I couldn't find anything on this topic yet.

**Comment by limeniye** [comment, marked as answer]
>Are events subscribed to in Xaml ever required to be unsubscribed? 

Events that subscribe to the same type do not need to be unsubscribed.

For example:
```C#
public MainPage()
{
    this.InitializeComponent();
    
    // in this case you don't need to unsubscribe the Loaded event, because the subscription is set to the same object.
    this.Loaded += OnMainPageLoaded;
    
    // in this case you don't need to unsubscribe the Tapped event, because if the MainPage removes, 
    // the MyTextBox removes also, because MyTextBox is the child of MainPage
    MyTextBox.Tapped += OnMyTextBoxTapped;
}
```

More information [there](https://stackoverflow.com/a/39100991/16493575)

>Are events subscribed to in code-behind ever required to be unsubscribed?

If you subscribe to an event from another type -- yes, you need an unsubscribe.
For example. If you have been worked with the **behaviors**, you have used the [OnAttached()](https://learn.microsoft.com/en-us/dotnet/api/microsoft.toolkit.uwp.ui.behaviors.behaviorbase-1.onattached?view=win-comm-toolkit-dotnet-7.1#microsoft-toolkit-uwp-ui-behaviors-behaviorbase-1-onattached) and [OnDetaching()](https://learn.microsoft.com/en-us/dotnet/api/microsoft.toolkit.uwp.ui.behaviors.behaviorbase-1.ondetaching?view=win-comm-toolkit-dotnet-7.1#microsoft-toolkit-uwp-ui-behaviors-behaviorbase-1-ondetaching) methods. That's needed to subscribe and unsubscribe your control events.
In [this example](https://github.com/Keboo/MaterialDesignInXaml.Examples/blob/ea3a5fa967b849515417cbf23a35a20bb7a53e05/TextBox/TextBox.SelectAllOnDoubleClick/SelectAllBehavior.cs#L14)  you can see that subscription to MouseDoubleClick of another control will be unsubscribed when that behavior will lose the dependency of that UI element and WinUI automaticly will execute the OnDetaching() method.

>Is there anything else here to be cautious of?

If in some case your MyTextBox can live longer than MainPage, I think this can happen if you move your element from one page to another (SecondPage for example), in this case you will definitely need an unsubscribe.

Also I want to say that there are ready-made implementations of the types for some cases.
For example, you can see how the Uno platform team is using the [SerialDisposable](https://github.com/unoplatform/uno/blob/7279060935b021bc4f25941308fc36c761764add/src/Uno.UI/DirectUI/IterableWrappedCollection.cs#L58C1-L59)

**Reply by abenadar** [reply]
Thanks a lot for these explanations. 
Looks like i have indeed one or two events that should be properly unsubscribed, and many unsubscribed though unnecessary.
Unfortunately it doesn't seem this has any impact on the memory leaks.
